### ISSUE TITLE
Batching of resolvers / data fetching

### ISSUE THREAD

**Issue by smatting** [morpheusgraphql/morpheus-graphql#348]
```graphql
type Deity {
  name: String!,
  children: [Deity]
}

type Query {
  deities(q: String) : [Deity]
}
```
Is it possible to implement resolvers that fetch all `children` for all the deities in one batch? Let's assume that you can resolve the list of all deities in one call, too.

Currently I only know how to write a lazy resolver for `children` which would be called for every `Deity` found in the resolver of `deities(q: String)`.

Could it by done by customizing the Monad `m` maybe?

**Comment by nalchevanidze** [comment]
@smatting sorry did not got. can you give me an example?

**Comment by nalchevanidze** [comment]
```gql
type Query {
  deities(q: String) : [Deity]
}
```

```hs
data Deity = Deity
  { name :: Text   
  , children   :: [Deity]
  } deriving (Generic,GQLType)

importGQLDocumentWithNamespace "schema.gql"

rootResolver :: GQLRootResolver IO () Query Undefined Undefined
rootResolver =
  GQLRootResolver
    {
      queryResolver = Query {queryDeities},
      mutationResolver = Undefined,
      subscriptionResolver = Undefined
    }
  where
    queryDeity QueryDeityArgs {queryDeityArgsName} = pure [ Deity {name ="", children=[...Deity...] },....]
```
do you mean this?

**Comment by smatting** [comment]
Let's assume that I have to fetch the `children` from a separate database `B`. Let's assume the resolver for `deities(q: String)` fetches 1000 deities from database `A` with one query, then to resolve `children` I don't want to make 1000 queries to database `B`.

**Comment by nalchevanidze** [comment]
```hs
data DBDeity = DBDeity { 
    dbID :: ID,
    dbName :: Name
}

dbDeities :: IO [DBDeity]
dbDeities = ....

dbPlacesByVisitors :: [ID] -> IO (Map ID Place)
dbPlacesByVisitors ids = ....

buildDeity ::   Map ID Place -> DBDeity -> Deity
buildDeity pls DBDeity{ id , dbName }  =  Diety {   
      places = maybe [] <$> lookup id pls,
      name = dbName 
}

deitiesResolver :: IORes e [Deity]
deitiesResolver = lift $ do
    dts <- dbDeities
    pls <- dbPlacesByVisitors (map dbID dts)
    pure $ map (buildDeity pls) dts
```

on way is that i think.

**Comment by nalchevanidze** [comment]
but then you loose laziness. you i ask places at db even if query does not needs it. may there is an automated way how they may batched, but i don't  know yet.

**Comment by nalchevanidze** [comment]
> Could it by done by customizing the Monad m maybe?

monad that bundles all your db requests in one request could be a solution. i am not quit familiar with [Haxl](https://github.com/facebook/Haxl) but may it could help.

one more thing what i'm thinking is this #258 proposal of @theobat, he is writing library that generates sql queries from `schema + request`.

**Comment by dandoh** [comment]
@smatting I believe we have complete control over this.
We can make a Deity resolver:
```haskell
deityResolver :: (DeityData, Maybe [PlaceData]) -> ResolverM Deity
deityResolver (deity, maybePlaces) = 
    Deity {..., placeVisited = placeVisitedResolver, ...}
  where 
    placeVisitedResolver = do 
        places <- case maybePlaces of 
            Just ps -> ps
            _ -> getDBDeityPlaces deity 
        mapM placeResolver places    
```
Then we know we can prefetch places for deities wherever we see fits:
```haskell
getDBDeitiesWithPlaces :: IO [(DeityData, PlaceData)]
getDBDeitiesWithPlaces = ...

deitiesResolver :: ResolverM [Deity]
deitiesResolver = do 
    deitiesWithPlaces <- fmap (mapSnd Just) getDBDeitiesWithPlaces
    mapM deityResolver deitiesWithPlaces
```
No more N + 1 query.

In terms of getting additional query information (such as visitedPlaces above), https://github.com/tomjaguarpaw/haskell-opaley is a good choice.

**Comment by smatting** [comment]
Thanks @dandoh! That's a cool trick to make a add optional prefetching like `deityResolver`, but `deitiesWithPlaces` is still loaded in every request, right? Your example solves the N+1 selects problem similar to @nalchevanidze 's example, but also sacrifices laziness of `placesVisited` (it's fetched even if not requested). Would be cool to have both:

1. `placesVisited` is resolved lazily (only fetched when requested)
2. No N + 1 selects

**Comment by smatting** [comment]
I accidentaly closed this issue. Re-opening.

**Comment by dandoh** [comment]
> 1. `placesVisited` is resolved lazily (places are fetched only when requested)
> 2. No N + 1 selects

That would be really cool. 
@nalchevanidze  What is the state of #258 ? 
Is it possible to make an instance:
```haskell
instance (MonadReader AST m) => Interpreter (Text -> m Text) m e where
  interpreter = ...
```
Then with access to query/mutation AST we can decide to prefetch or not.

**Comment by nalchevanidze** [comment]
actally i am planning to add new feature `getContext` which will you give access to current state of resolving. 

```hs
deitiesResolver :: IORes e [Deity]
deitiesResolver = do
    Context { selection } <- getContext
    dts <- lift dbDeities
    pls <- lift dbPlacesByVisitors selection (map dbID dts)
    pure $ map (buildDeity pls) dts
```
disadvantage is that resolver will depend on internal `AST`.

**Comment by nalchevanidze** [comment]
> @dandoh, actally i am planning to add new feature getContext which will you give access to current internal state of resolving.
see, #372

**Comment by AnthonySuper** [comment]
Currently we can sort of do preloading via the internal context. However, as you said, the internal context should be considered "unsafe" as it exposes internal details. Are there any plans for a more public version of this API? The basic case would be to have some sort of recursive query type that told you the fields requested and the arguments passed, so you could do preloading. Unfortunately I am having a lot of trouble coming up with a typesafe API to do so, but in principle one should be possible, I think?

**Comment by nalchevanidze** [comment]
@AnthonySuper what if we define function that can search in sub selections.

```hs
path :: Text -> Resolver Bool
-- so you can ask
somRes = do
    prefetch <- path "field1.field2"
    ...
```

**Comment by AnthonySuper** [comment]
That would be a great start! Eventually I'd love to get something where we can get arguments to prefetch too, but messing around with that on my end has given me a nontrivial amount of trouble in figuring out how the types would work.

**Comment by Herlevsen** [comment]
The way some other graphql libraries solves batching, is by using the concept of a dataloader. The dataloader can batch requests to the same resources, and also handles caching to avoid unnecessary lookups. Facebooks implementation, for javascript/node is located here https://github.com/graphql/dataloader. They actually mention Haxl, and it seems to be the exact thing it was created for

**Comment by AnthonySuper** [comment]
Haxl definitely seems to be a good option. I'm currently toying around with a simpler version that uses laziness (and `unsafeInterleaveIO`) to rewrite the actions in the graph, which seems to work out okay—I'll probably write a blog post about it if it might help out other people.

**Comment by nalchevanidze** [comment]
@AnthonySuper great idea :) i would love  to read it

**Comment by theobat** [comment]
> @AnthonySuper what if we define function that can search in sub selections.
> 
> ```haskell
> path :: Text -> Resolver o m Bool
> -- so you can ask
> somRes = do
>     (shouldPrefetch :: Bool) <- path "field1.field2"
>     ...
> ```

Regarding this, I believe we should simply implement a typeclass instance of a Tree ([such as this one for example](https://hackage.haskell.org/package/hxt-9.3.1.18/docs/Data-Tree-Class.html)), this would likely simplify a few internal operations (print the AST back to a doc etc) and give simple (read canonical) utility functions to manipulate the query's AST while retaining flexibility for the concrete (internal) representation. I can probably try a PR for this.

Also it enables not exporting the concrete Rep while giving freedom for the end-user (which was your concern I believe @nalchevanidze)

**Comment by nalchevanidze** [comment]
@theobat 
what incase of `UnionSelection`?

https://github.com/morpheusgraphql/morpheus-graphql/blob/68ac1a95d1238db5c23cd18f6ec6b70567c11ff1/morpheus-graphql-core/src/Data/Morpheus/Types/SelectionTree.hs#L35-L37

we may should add:
- `isUnion :: node -> Bool` 
- `getUnionChildrens :: TypeName -> node -> [node]` 


ps. i think we should rename `getChildrenList ` -> `getChildrens`.

**Comment by theobat** [comment]
Yep I'll change the name. As far as the union is concerned, is it actually separated in the selection set ? or mixed ? I think it'd be simpler to just give a ```getTypeName :: node -> TypeName``` kind of operation though...

**Comment by nalchevanidze** [comment]
sorry. but union types in selection does not work that way.

**Comment by nalchevanidze** [comment]
but we can  define virtual type`SelectionNode' that can support it. another concern is.

```hs 
getChildren ::  node -> [(Maybe ConditionTypeName, node)]
```

**Comment by nalchevanidze** [comment]
or 
```hs 
getChildren ::  node -> Either [(ConditionTypeName, node)] [node]
```
or just

**Comment by theobat** [comment]
I made a typo I meant ```getTypeName :: node -> Maybe TypeName```. If the use case described is this:
```gql
{
  search(text: "an") {
    __typename
    ... on Human {
      name
      height
    }
    ... on Droid {
      name
      primaryFunction
    }
    ... on Starship {
      name
      length
    }
  }
}
```
Then it should work, and I find it "more general" in the sense that union type or not, every graphql node has a type it inhabits

**Comment by nalchevanidze** [comment]
i meant union types in Morpheus GraphQL internaly union selection is not  represented that way.

we you should wrap it as.

```
newtype SelectionNode = SelectionNode {  typeName:: TypeName, selection :: SelectionNode  }
```

and  you need additionally the schema to get selection Type.

**Comment by nalchevanidze** [comment]
another solution can be:

```hs
getChildren ::  Maybe TypeName ->  node -> [node]
```
where Nothing means give me selection without any condition

**Comment by dandoh** [comment]
@nalchevanidze @theobat 
And discussed with @nalchevanidze, 

How hard is it to make another intepretation: resolver by field instead of by type as that of mu-haskell, on top of what we had?


```
User {
  name: String
  dog: Dog
}

Dog {
  name: string
}
```

From this we can derive the Selection type family:

```haskell
data Selection String = Bool -- For any scalar
data Selection User = { name : Selection String, dog : Selection Dog }
data Selection Dog = { name : Selection String }
```

Then each resolver will have the coressponding Selection as an argument.
https://github.com/higherkindness/mu-haskell/issues/190

**Comment by nalchevanidze** [comment]
@dandoh should you show me what you mean? please give a concrete example like @smatting

**Comment by nalchevanidze** [comment]
@smatting @dandoh @AnthonySuper @Herlevsen  i think i will provide batching api similar to https://github.com/graphql/dataloader

**Comment by russellmcc** [comment]
Just wanted to chime in and say that I used [fraxl](https://github.com/ElvishJerricco/fraxl), a clone of Haxl with more general types, to implement batching in my morpheus project.  This worked really well for the purpose, and fulfills both requirements listed above (queries happen lazily, no N+1 problem).  

To make it work, basically you define all your possible db queries in a GADT, then define a fetcher function that executes the queries.  The fraxl library then calls the fetcher with groups of queries that can be batched together.  fraxl uses the applicative `<*>` to determine what group of queries can be batched together.  For my (simple) program it seemed to give good results without much tweaking.

I'm not ready to share the whole project at the moment but I can go in to more detail on how this works if people are interested.  It also supports optional caching which is a nice speed up if you happen to re-query anything.

**Comment by nalchevanidze** [comment]
@russellmcc cool. can you add a example code to this project?

**Comment by nalchevanidze** [comment]
I know you are not ready to share your project, but if you come up with some basic example could help :)

**Comment by russellmcc** [comment]
Can do!  Look for a PR sometime over the next week

**Comment by zhujinxuan** [comment]
There is also something related about batching and deferring evaluation.  Suppose we have a chained relation in the graphql, we may want to defer the evaluation.  

```
type User {
   name: String!
   articles: [Article!]!
}

type Article {
  content: String!,
  writer: User
}

```

I think that we can change the generated datatype as 
```
data UserM m =    UserM {
   name: m Text,
   articles:  m [ArticleM m]
}

data ArticleM m = ArticleM {
   content: m Text,
   writer: m (UserM m)
}
```

In this way, we can ensure that the evaluation of one field does not necessarily trigger the evaluation of other field.  For batching, it seems possible to have a dataloader in `m` as a ReadT monad or something alike to solve the problem.

**Comment by nalchevanidze** [comment]
@smatting your original request will be addressed in #786 with named resolvers
